### PR TITLE
8277976: Break up SEQUENCE in X509Certificate::getSubjectAlternativeNames and X509Certificate::getIssuerAlternativeNames in otherName

### PR DESCRIPTION
**Pull request by wangweij**
The enhancement adds two extra items in the `getSubjectAlternativeNames()` output for an OtherName.

It also fix several errors:
1. In `OtherName.java`, `nameValue` should be the value inside `CONTEXT [0]` without the tag and length bytes.
2. The argument in constructor `extClass.getConstructor(Object.class)` is suspicious. Maybe it meant `byte[]`.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed
- [x] Change requires a CSR request to be approved

### Issues
 * [JDK-8277976](https://bugs.openjdk.java.net/browse/JDK-8277976): Break up SEQUENCE in X509Certificate::getSubjectAlternativeNames and X509Certificate::getIssuerAlternativeNames in otherName
 * [JDK-6776681](https://bugs.openjdk.java.net/browse/JDK-6776681): Invalid encoding of an OtherName in X509Certificate.getAlternativeNames()
 * [JDK-8281606](https://bugs.openjdk.java.net/browse/JDK-8281606): Break up SEQUENCE in X509Certificate::getSubjectAlternativeNames and X509Certificate::getIssuerAlternativeNames in otherName (**CSR**)

### Reviewers
 * [Sean Mullan](https://openjdk.java.net/census#mullan) (@seanjmullan - **Reviewer**) ⚠️ Review applies to bca0b6d3a54d14e747aba0b8bcb65bef2092d3c6

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/7167/head:pull/7167` \
`$ git checkout pull/7167`

Update a local copy of the PR: \
`$ git checkout pull/7167` \
`$ git pull https://git.openjdk.java.net/jdk pull/7167/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 7167`

View PR using the GUI difftool: \
`$ git pr show -t 7167`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/7167.diff">https://git.openjdk.java.net/jdk/pull/7167.diff</a>

</details>

### CODE REVIEW

File: test/jdk/sun/security/x509/OtherName/Parse.java [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.security.cert.X509Certificate;
+import java.util.Arrays;
+import java.util.Date;
+
+public class Parse {
+
+    public static class MyDNSName extends DNSName {
```
Comment (seanjmullan) [inline test/jdk/sun/security/x509/OtherName/Parse.java:54]: What's the purpose of the MyDNSName class? You don't seem to do anything with it later in the test.

Comment (wangweij) [reply test/jdk/sun/security/x509/OtherName/Parse.java:54]: It's a kind of otherName (the assigned OID is 1.2.3.6) I used later. The `OtherName` class tries to parse the name inside (see the private `getGNI()` method) and it is used in `equals()`. I just want to confirm it works.

File: src/java.base/share/classes/java/security/cert/X509Certificate.java [hunk 2]
```diff
@@ -581,9 +585,14 @@ public X500Principal getSubjectX500Principal() {
      * in the form "a1:a2:...:a8", where a1-a8 are hexadecimal values
      * representing the eight 16-bit pieces of the address. OID names are
      * returned as {@code String}s represented as a series of nonnegative
-     * integers separated by periods. And directory names (distinguished names)
+     * integers separated by periods. Directory names (distinguished names)
      * are returned in <a href="http://www.ietf.org/rfc/rfc2253.txt">
-     * RFC 2253</a> string format. No standard string format is
+     * RFC 2253</a> string format. Othernames are returned as a byte array
```
[the remark this answers is not in the document]
Comment (wangweij) [reply src/java.base/share/classes/java/security/cert/X509Certificate.java:590]: OK.

**Comment by wangweij** [comment]
> Looks good, but I think a CSR should also be filed.

Sure, I'll write one now. I've added the `csr` label so that I will not forget about it. Just want to delay the writing after we agree on the text.

**Comment by seanjmullan** [comment]
One issue with this change is that there is no guarantee when or if a 3rd party provider will support these additional elements for OtherNames. In a worst case scenario, code may get an IOOBException if they assume there will be a 3rd or fourth element and there isn't. There is no perfect solution for this. I would suggest rewording the new text with a "may", and adding an @implNote stating that the SUN provider supports this feature. Something like:

```
      * otherNames are returned as a byte array
      * containing the ASN.1 DER encoded form of the name, and
      * may also return a third entry in the list containing its {@code type-id} in string
      * format, and a fourth entry containing the ASN.1 DER encoding of
      * its {@code value} without the context-specific constructed tag
      * with number 0.
      *
      * @implNote The JDK SUN provider supports the third and fourth otherName elements. 
```

File: src/java.base/share/classes/java/security/cert/X509Certificate.java [hunk 3]
```diff
@@ -581,9 +585,14 @@ public X500Principal getSubjectX500Principal() {
...
      * returned as {@code String}s represented as a series of nonnegative
-     * integers separated by periods. And directory names (distinguished names)
+     * integers separated by periods. Directory names (distinguished names)
      * are returned in <a href="http://www.ietf.org/rfc/rfc2253.txt">
-     * RFC 2253</a> string format. No standard string format is
+     * RFC 2253</a> string format. otherNames are returned as a byte array
+     * containing the ASN.1 DER encoded form of the name, and may also return
+     * a third entry in the list containing its {@code type-id} in string
+     * format, and a fourth entry containing the ASN.1 DER encoding of
+     * its {@code value} without the context-specific constructed tag
```
Comment (seanjmullan) [inline src/java.base/share/classes/java/security/cert/X509Certificate.java:594]: To be very specific, I would add that the 4th entry is a byte array, ex: "and a fourth entry as a byte array containing the ASN.1 DER encoding ..."

Comment (wangweij) [reply src/java.base/share/classes/java/security/cert/X509Certificate.java]: Good idea. Fixed and updated the CSR as well.

**Review by seanjmullan — APPROVED** [review]

**Comment by michael-o** [comment]
@wangweij I would highly recommend to address this ticket first: https://bugs.openjdk.java.net/browse/JDK-6776681. It is partially related.

File: src/java.base/share/classes/sun/security/x509/OtherName.java [hunk 4]
```diff
@@ -89,8 +89,11 @@ public OtherName(DerValue derValue) throws IOException {
         DerInputStream in = derValue.toDerInputStream();
 
         oid = in.getOID();
-        DerValue val = in.getDerValue();
-        nameValue = val.toByteArray();
+        DerValue derValue1 = in.getDerValue();
+        if (!derValue1.isContextSpecific((byte)0)) {
```
Comment (michael-o) [inline src/java.base/share/classes/sun/security/x509/OtherName.java:93]: Where do you check for constructed (0x20)?

File: test/jdk/sun/security/x509/OtherName/Parse.java [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+        CertAndKeyGen g = new CertAndKeyGen("Ed25519", "Ed25519");
+        g.generate(-1);
+        X509Certificate x = g.getSelfCertificate(new X500Name("CN=ME"),
+                new Date(),
+                100000,
+                exts);
+
+        int found = 0;
+        for (var san : x.getSubjectAlternativeNames()) {
+            if (san.get(2).equals("1.2.3.5")
```
Comment (michael-o) [inline test/jdk/sun/security/x509/OtherName/Parse.java:86]: Would it make sense to explicitly test length for at least 4?

**Comment by michael-o** [comment]
@wangweij Any chance to evaluate this comment from me: https://bugs.openjdk.java.net/browse/JDK-8277976?focusedCommentId=14462729&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-14462729?
Automatic mapping of the value if it is a ASN.1 simple type to a Java type like done for the remaining general names? At least strings would be straight forward and cover many cases, e.g., MS UPN (1.3.6.1.4.1.311.20.2.3).

**Comment by wangweij** [comment]
> @wangweij I would highly recommend to address this ticket first: https://bugs.openjdk.java.net/browse/JDK-6776681. It is partially related.

I noticed the problem and have fixed it in this code change. `nameValue` is the bytes inside.

File: test/jdk/sun/security/x509/OtherName/Parse.java (hunk 5 above)
Comment (wangweij) [reply test/jdk/sun/security/x509/OtherName/Parse.java:86]: Sure.

File: src/java.base/share/classes/sun/security/x509/OtherName.java (hunk 4 above)
Comment (wangweij) [reply src/java.base/share/classes/sun/security/x509/OtherName.java:93]: Oops, I should add an `isConstructed()` check.

**Comment by wangweij** [comment]
> @wangweij Any chance to evaluate this comment from me: https://bugs.openjdk.java.net/browse/JDK-8277976?focusedCommentId=14462729&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-14462729?
> Automatic mapping of the value if it is a ASN.1 simple type to a Java type like done for the remaining general names? At least strings would be straight forward and cover many cases, e.g., MS UPN (1.3.6.1.4.1.311.20.2.3).

I'll think about it. Strings (except for OCTET STRING and BIT STRING) are OK.

File: test/jdk/sun/security/x509/OtherName/Parse.java [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+        CertAndKeyGen g = new CertAndKeyGen("Ed25519", "Ed25519");
+        g.generate(-1);
+        X509Certificate x = g.getSelfCertificate(new X500Name("CN=ME"),
+                new Date(),
+                100000,
+                exts);
+
+        int found = 0;
+        for (var san : x.getSubjectAlternativeNames()) {
+            if (san.size() == 4) {
```
Comment (michael-o) [inline test/jdk/sun/security/x509/OtherName/Parse.java:89]: Make it `>=` if this is going to change in the future..

Comment (wangweij) [reply test/jdk/sun/security/x509/OtherName/Parse.java:89]: OK.

File: src/java.base/share/classes/sun/security/x509/OtherName.java [hunk 7]
```diff
@@ -89,8 +89,11 @@ public OtherName(DerValue derValue) throws IOException {
         DerInputStream in = derValue.toDerInputStream();
 
         oid = in.getOID();
-        DerValue val = in.getDerValue();
-        nameValue = val.toByteArray();
+        DerValue derValue1 = in.getDerValue();
+        if (!derValue1.isContextSpecific((byte)0) || !derValue1.isConstructed()) {
```
Comment (michael-o) [inline src/java.base/share/classes/sun/security/x509/OtherName.java:93]: It might be purely a matter of taste, but isn't `!(isCSTag0 && isConst)` easier to read?

Comment (wangweij) [reply src/java.base/share/classes/sun/security/x509/OtherName.java:93]: I have difficulty describing `!(a && b)`. There is no parentheses in human language and '!' has higher order than '&&'.

I thought about completely reverse the block but that means everything after the throw will be inside a block and I don't want to move so many lines.

File: src/java.base/share/classes/sun/security/x509/X509CertImpl.java [hunk 8]
```diff
@@ -1584,6 +1584,17 @@ public int getBasicConstraints() {
...
                 }
                 nameEntry.add(derOut.toByteArray());
+                if (name.getType() == GeneralNameInterface.NAME_ANY
+                        && name instanceof OtherName oname) {
+                    nameEntry.add(oname.getOID().toString());
+                    byte[] nameValue = oname.getNameValue();
+                    try {
+                        String v = new DerValue(nameValue).getAsString();
+                        nameEntry.add(v == null ? nameValue : v);
+                    } catch (IOException ioe) {
```
Comment (michael-o) [inline src/java.base/share/classes/sun/security/x509/X509CertImpl.java:1594]: Attention, this catch block will hide invalid ASN.1 encoding of the other name element from:
* sun.security.util.DerValue.init(boolean, InputStream, boolean)
* sun.security.util.DerValue.getIA5String()

Other blocks throw:
```
throw new CertificateException("Unable to parse DER value of SAN:otherName", ioe);
```

Do you really intend to hide an encoding error int the cert from the user?

File: test/jdk/sun/security/x509/OtherName/Parse.java (hunk 6 above)
Comment (michael-o) [reply test/jdk/sun/security/x509/OtherName/Parse.java:89]: Right, let's leave as-is.

File: src/java.base/share/classes/sun/security/x509/X509CertImpl.java (hunk 8 above)
Comment (wangweij) [reply src/java.base/share/classes/sun/security/x509/X509CertImpl.java:1594]: Up to debate. Other blocks in `makeAltNames` throw `RuntimeException`.

File: src/java.base/share/classes/sun/security/x509/OtherName.java (hunk 7 above)
Comment (michael-o) [reply src/java.base/share/classes/sun/security/x509/OtherName.java:93]: My wording for the &&: If the tag is not a constructed and context-specific tag number 0, then an exception is thrown. The parens denote that both conditions need to apply:
```
!(isCSTag0 && isConst)

true, true = !(true && true) = !true = false
true, false = !(true && false) = !false = true
false, true = !(false && true) = !false = true
false, false = !(false && false) = !false = true


!isCSTag0 || !isConst

true, true = !true || !true = false || false = false
true, false = !true || !false = false || true = true
false, true = !false || !true = true || false = true
false, false = !false || !false = true || true = true
```

Comment (wangweij) [reply src/java.base/share/classes/sun/security/x509/OtherName.java:93]: If you really like it, I'll write
```
        if (derValue1.isContextSpecific((byte) 0) && derValue1.isConstructed()) {
            nameValue = derValue1.data.toByteArray();
            gni = getGNI(oid, nameValue);
            if (gni != null) {
                name = gni.toString();
            } else {
                name = "Unrecognized ObjectIdentifier: " + oid.toString();
            }
        } else {
            throw new IOException("value is not [0]");
        }
```

**Comment by wangweij** [comment]
> Are you going to address this separately or document to be implicitly fixed by this PR?

Normally we close the other one as a duplicate. I'll do it now.

**Comment by michael-o** [comment]
> > Are you going to address this separately or document to be implicitly fixed by this PR?
> 
> Normally we close the other one as a duplicate. I'll do it now.

Well, technically it is not a duplicate since both tickets address different problems...

File: src/java.base/share/classes/sun/security/x509/X509CertImpl.java (hunk 8 above)
Comment (wangweij) [reply src/java.base/share/classes/sun/security/x509/X509CertImpl.java:1594]: But in this case, we still have a place to provide the raw bytes. Maybe that's better? Or you'd rather be sure that one particular otherName should always have a string there and there's no need to do an `instanceof` check?

**Comment by michael-o** [comment]
> New commit pushed with a spec change and CSR is now in draft mode. Something worth mentioning:
> 
>     1. I explicitly added "valid UTF8String..." to match the `catch (IOException)` line in the implementation, so if there's anything wrong parsing the value as a string then the byte array will still be there.

Good.
 
>     2. I feel a little uneasy of the new `if` and `otherwise` words inside parentheses, especially the second one which seems out of nowhere. Please suggest better wording if possible.

What about?

    and a fourth entry as either a string if {@code value} inside
    is a valid ASN.1 string (any of UTF8String, PrintableString, T61String, IA5String,
    UniversalString, BMPString, or GeneralString) or otherwise a byte array containing
    the ASN.1 DER encoding of {@code value} without the context-specific
    constructed tag with number 0. 

>     3. The string types I listed is only what we supported inside JDK and not all "restricted character string types". Is it appropriate to list them out in a specification? If not, shall I just say "a valid character string"? This brings more explanation to what "valid" means.

Yes, this makes it perfectly clear what is supported of out of the box.

File: src/java.base/share/classes/sun/security/x509/X509CertImpl.java (hunk 8 above)
Comment (michael-o) [reply src/java.base/share/classes/sun/security/x509/X509CertImpl.java:1594]: I have thought about this actually. Now the parse is free of any semantics, which is naive. Actually, you need a list of wellknown OIDs to know the target tag type to perform the conversion. E.g., you know that MS UPN must be UTF8String, if not this is an error. If you don't know the OID, don't touch it.
OpenSSL knows the semantics and decodes it, otherwise don't touch it and leave it: https://github.com/openssl/openssl/blob/317acac5cc0a2cb31bc4b91353c2b752a3989d8a/crypto/x509/v3_san.c#L113-L120

Comment (michael-o) [reply src/java.base/share/classes/sun/security/x509/X509CertImpl.java:1594]: Maybe adopt the list of OpenSSL, otherwise return byte array?

**Comment by wangweij** [comment]
So my current opinion is to make it as vague as possible (Ex: `a valid supported character string`) so people always remember to check `instanceof String` first.

**Comment by michael-o** [comment]
> So my current opinion is to make it as vague as possible (Ex: `a valid supported character string`) so people always remember to check `instanceof String` first.

I agree with that.

**Comment by wangweij** [comment]
> I don't expect any new ASN.1 string types to be added in the future, but of someone decides to create a public ASN.1 

I've seen new string types that need 2 bytes tag, but don't know if they are used anywhere.

**Comment by michael-o** [comment]
> > I don't expect any new ASN.1 string types to be added in the future, but of someone decides to create a public ASN.1
> 
> I've seen new string types that need 2 bytes tag, but don't know if they are used anywhere.

Oh, can you reference? 

> Also, there are existing types we don't support also, like TeletexString.

That is dead, isn't it?

File: src/java.base/share/classes/java/security/cert/X509Certificate.java [hunk 9]
```diff
@@ -581,9 +585,16 @@ public X500Principal getSubjectX500Principal() {
...
+     * containing the ASN.1 DER encoded form of the name, and may also return
+     * a third entry in the list containing its {@code type-id} in string
+     * format, and a fourth entry as either a string (if {@code value} inside
+     * is a valid UTF8String, PrintableString, T61String, IA5String,
+     * UniversalString, BMPString, or GeneralString) or a byte array containing
+     * the ASN.1 DER encoding of {@code value} without the context-specific
+     * constructed tag with number 0 (otherwise). No standard string format is
      * defined for otherNames, X.400 names, EDI party names, or any
      * other type of names. They are returned as byte arrays
      * containing the ASN.1 DER encoded form of the name.
```
Comment (seanjmullan) [inline src/java.base/share/classes/java/security/cert/X509Certificate.java:600]: I would move this sentence up before "otherNames ..." and remove the "otherNames" from it and the end of the sentence. So basically:

>      * No standard string format is
>      * defined for X.400 names or EDI party names. They are returned as byte arrays
>      * containing the ASN.1 DER encoded form of the name.
>      * otherNames are returned as a byte array ...

**Comment by wangweij** [comment]
> Oh, can you reference?

Sorry, memory is incorrect, those are date time types (in https://www.oss.com/asn1/resources/reference/ASN.1-Reference-Card-format-USA.pdf). But here they do promote a VisibleString that we don't support.

**Comment by wangweij** [comment]
New commit pushed. For the openssl style suggestion, I think its major benefit is to provide a string format of the type (like `"othername: UPN:"`). In the `default` block, it still extracts either the IA5String or the UTF8String. I think it's not worth penalizing people putting an IA5String into a UPN. Right?

**Comment by michael-o** [comment]
> New commit pushed. For the openssl style suggestion, I think its major benefit is to provide a string format of the type (like `"othername: UPN:"`). In the `default` block, it still extracts either the IA5String or the UTF8String. I think it's not worth penalizing people putting an IA5String into a UPN. Right?

UPN is a DirectoryString from AD which is UTF-8 encoded. The default case from OPENSSL tries to cover OIDs it does not know. UPN it does know and its semantics is always UTF8String. Everything else is wrong. I5AString is rather email which is another general name. We have now two options: add well known and decode accordingly or do yhe default case. From a user's perspective the encoding is rather irrelevant because he wants Java types.

**Comment by wangweij** [comment]
> > New commit pushed. For the openssl style suggestion, I think its major benefit is to provide a string format of the type (like `"othername: UPN:"`). In the `default` block, it still extracts either the IA5String or the UTF8String. I think it's not worth penalizing people putting an IA5String into a UPN. Right?
> 
> UPN is a DirectoryString from AD which is UTF-8 encoded. The default case from OPENSSL tries to cover OIDs it does not know. UPN it does know and its semantics is always UTF8String. Everything else is wrong. I5AString is rather email which is another general name. We have now two options: add well known and decode accordingly or do yhe default case. From a user's perspective the encoding is rather irrelevant because he wants Java types.

I'll pick the 2nd option now.

File: src/java.base/share/classes/java/security/cert/X509Certificate.java [hunk 10]
```diff
@@ -581,12 +585,17 @@ public X500Principal getSubjectX500Principal() {
...
-     * other type of names. They are returned as byte arrays
-     * containing the ASN.1 DER encoded form of the name.
+     * RFC 2253</a> string format. No standard string format is defined for
+     * X.400 names or EDI party names. They are returned as byte arrays
+     * containing the ASN.1 DER encoded form of the name. otherNames are also
+     * returned as byte arrays containing the ASN.1 DER encoded form of the
+     * name. There may be a third entry in the list for their {@code type-id}
+     * in string form, and a fourth entry for their {@code value} as either
+     * a string (if the value is a valid supported character string)
+     * or (otherwise) a byte array containing the ASN.1 DER encoded from of
```
Comment (seanjmullan) [inline src/java.base/share/classes/java/security/cert/X509Certificate.java:597]: typo: s/from/form/

File: src/java.base/share/classes/java/security/cert/X509Certificate.java [hunk 11]
```diff
@@ -599,6 +608,9 @@ public X500Principal getSubjectX500Principal() {
      * and it provides a default implementation. Subclasses
      * should override this method with a correct implementation.
      *
+     * @implNote The JDK SUN provider supports the third and fourth
+     * otherName elements.
```
Comment (seanjmullan) [inline src/java.base/share/classes/java/security/cert/X509Certificate.java:612]: I would probably use "entries" instead of "elements" to be consistent with the previous wording.

File: test/jdk/sun/security/x509/OtherName/Parse.java [hunk 12]
```diff
@@ -0,0 +1,102 @@
...
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+/*
+ * @test
+ * @bug 8277976
+ * @summary Break up SEQUENCE in X509Certiticate::getSubjectAlternativeNames
```
Comment (seanjmullan) [inline test/jdk/sun/security/x509/OtherName/Parse.java:27]: Typo: s/X509Certiticate/X509Certificate/ 
(also on line 28)

File: src/java.base/share/classes/java/security/cert/X509Certificate.java [hunk 13]
```diff
@@ -581,12 +585,17 @@ public X500Principal getSubjectX500Principal() {
...
-     * RFC 2253</a> string format. No standard string format is
-     * defined for otherNames, X.400 names, EDI party names, or any
-     * other type of names. They are returned as byte arrays
-     * containing the ASN.1 DER encoded form of the name.
+     * RFC 2253</a> string format. No standard string format is defined for
+     * X.400 names or EDI party names. They are returned as byte arrays
+     * containing the ASN.1 DER encoded form of the name. otherNames are also
+     * returned as byte arrays containing the ASN.1 DER encoded form of the
+     * name. There may be a third entry in the list for their {@code type-id}
+     * in string form, and a fourth entry for their {@code value} as either
```
[the remark this answers is not in the document]
Comment (wangweij) [reply src/java.base/share/classes/java/security/cert/X509Certificate.java:595]: The whole paragraph uses plural forms like "IPv4 address names are" and "OID names are", so I also use "otherNames are" and "their". I admit I don't feel comfortable too.

File: src/java.base/share/classes/java/security/cert/X509Certificate.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (wangweij) [reply src/java.base/share/classes/java/security/cert/X509Certificate.java:597]: OK.

File: src/java.base/share/classes/java/security/cert/X509Certificate.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (wangweij) [reply src/java.base/share/classes/java/security/cert/X509Certificate.java:595]: I thought about "the", but it does not show the belong to relation.

File: test/jdk/sun/security/x509/OtherName/Parse.java (hunk 12 above)
Comment (wangweij) [reply test/jdk/sun/security/x509/OtherName/Parse.java:27]: On no.

File: src/java.base/share/classes/java/security/cert/X509Certificate.java (hunk 11 above)
Comment (wangweij) [reply src/java.base/share/classes/java/security/cert/X509Certificate.java:612]: Good.

File: src/java.base/share/classes/java/security/cert/X509Certificate.java (hunk 10 above)
Comment (wangweij) [reply src/java.base/share/classes/java/security/cert/X509Certificate.java:597]: Oops.

File: src/java.base/share/classes/java/security/cert/X509Certificate.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (seanjmullan) [reply src/java.base/share/classes/java/security/cert/X509Certificate.java:595]: What you can do is reword the sentence by including `otherName` in it, ex: "A third entry may also be present in the list containing the {@code type-id} of the otherName in string form, and a fourth entry containing the {@code value} of the otherName as either ...".

File: src/java.base/share/classes/sun/security/x509/X509CertImpl.java (hunk 8 above)
Comment (michael-o) [reply src/java.base/share/classes/sun/security/x509/X509CertImpl.java:1594]: @wangweij So you have selected to really swallow the exception here?

Comment (wangweij) [reply src/java.base/share/classes/sun/security/x509/X509CertImpl.java:1594]: Yes. I don't want to let the method fail. Since `instanceof String` should be called anyway the caller can decide how to fail.

Comment (michael-o) [reply src/java.base/share/classes/sun/security/x509/X509CertImpl.java:1594]: So you leave it to the user to read the ASN.1 value and fail if the encoding is incorrect instead of throwing a `Ceritificate..Exception` although other GeneralNames do?

Comment (wangweij) [reply src/java.base/share/classes/sun/security/x509/X509CertImpl.java:1594]: I want to minimize behavior change and I'm leaving them to notice there's no string there and fail.

Comment (wangweij) [reply src/java.base/share/classes/sun/security/x509/X509CertImpl.java:1594]: For other known names, you either return a parsed name or fail. For `otherName`, you already have the raw data and this further parsed name is a bonus. I just don't like a problem in getting the bonus to ruin the original benefit. Also, I think the caller will have to check the length and the type anyway so this will not be an extra burden. Besides, I always feel that `otherName` could be freely extended and the quality of its encoding might not be as guaranteed as the other ones. @seanjmullan, any comment here?

Too many "other" used and hopefully no one get confused.

Comment (seanjmullan) [reply src/java.base/share/classes/sun/security/x509/X509CertImpl.java:1594]: This seems like a reasonable balance between preserving existing behavior and still providing the data for the application to inspect. To help debugging, you could log the exception.

**Comment by wangweij** [comment]
> _Mailing list message from Michael StJohns...

I'll use your exception message on not a `[0]`.

For dealing with the error while the parsing of `nameValue`, we've discussed it and my last comment is at https://github.com/openjdk/jdk/pull/7167#discussion_r810155556. I'd rather not fail.

**Comment by wangweij** [comment]
I haven't logged the IOException. When this (otherName with a char string value but 4th entry is not string) happens, this is the only reason.
